### ISSUE TITLE
[Feature request] Add ability to "impersonate contract"

### ISSUE THREAD

**Issue by nlordell** [nomiclabs/hardhat#1355]
### Summary

`CREATE2` is great an allows contracts to be deployed at deterministic addresses. The mechanism for determining the contract address, however, mixes in the contract bytecode hash. This makes it impossible to use a Waffle `MockContract`, for example, in order to mock interactions with contracts at `CREATE2` addresses. If there were a way to set code at a specific address (this would rely on `ethereumjs-vm` supporting such a feature), then you could set the code at the `CREATE2` address to be a mock contract, and easily unit test interactions with mocked responses.

### Example

So, to illustrate where this could be useful, for example, consider a simple contract that interacts with a Uniswap V2 pair:
```solidity
contract Foo {
    function bar(IERC20 tokenA, IERC20 tokenB) external {
        IUniswapV2Pair pair = pairFor(tokenA, tokenB);
        // Do stuff with `pair`...
    }

    function pairFor(IERC20 tokenA, IERC20 tokenB) private returns (IUniswapV2Pair pair) {
        // compute `CREATE2` address ...
    }
}
```

Now, if we want to use a mocked pair instance, one way to do it currently is to use function overloading:
```solidity
contract Foo {
    // ...
    function pairFor(IERC20 tokenA, IERC20 tokenB) internal virtual returns (IUniswapV2Pair pair)
    // ...
}

contract FooTest is Foo {
    address private immutable mock;
    constructor(address mock_) {
        mock = mock_;
    }
    function pairFor(IERC20 tokenA, IERC20 tokenB) private override returns (IUniswapV2Pair pair) {
        return IUniswapV2Pair(mock);
    }
}
```

We would then use `FooTest` for unit testing our `Foo` contract, allowing us to mock `UniswapV2Pair` contract calls.

This however, is cumbersome, and this workaround isn't always possible (for example if the `CREATE2` address is being computed in a library).

IMO a better solution would be to allow setting code at a specific address (again, assuming that this is supported by `ethereumjs-vm`). This would allow us to write test code with mocked contracts without any code changes. Something like:
```ts
describe("Foo", () => {
  it("interacts with Uniswap pair", async () => {
    const [tokenA, tokenB] = ...
    const foo = await ethers.getContractFactory("Foo").then(factory => factory.deploy());
    const mockPair = await waffle.deployMockedContractAt(
      deployer,
      pairFor(tokenA, tokenB),
      UNISWAP_PAIR_ABI,
    );
    
    await mockPair.mock...
    await expect(foo.bar(tokenA, tokenB)).to.not.be.reverted;
  });
});
```

Again, this would rely on whether or not the Ethereum VM implementation allows this, but I think this could be a very powerful feature to complement "impersonate account".

### Other Uses

Another great use for impersonating account is mainnet forking. This can be powerful in those situations as well. Off the top of my head, I can imagine a situation where you would like to validate a fix to some code. Using a mainnet for, modified contract code could "replace" an existing mainnet contract, and a transaction can be replayed with the code modification to see how the modified behaviour would play out.

**Comment by fvictorio** [comment]
We want to add a `hardhat_setCode` RPC method for changing the code of a deployed contract, would that be enough or do you think something else would be needed to satisfy your use case?

**Comment by nlordell** [comment]
That would. Would `hardhat_setCode` allow setting code for any address or only addresses that already contain code?

**Comment by fvictorio** [comment]
> That would. Would `hardhat_setCode` allow setting code for any address or only addresses that already contain code?

That's something we discussed. We can do both, it's a question of what is better.

Allowing this for any address is more flexible, but limiting it to address with code would help catch errors where you think the address is a contract but it isn't.

Other options are:

- Allow in any address, but print a warning if the address didn't already have code.
- Disallow setting code for addresses without code, but accept an optional argument (like `{ force: true }`) that lets you do this.

I'm inclined to just accept any address and let it be, tbh.

**Comment by fvictorio** [comment]
Maybe the best option is to allow this for any address and to always print a message like:

```
Code in address 0x123... was updated. It had 123 bytes and now it has 234 bytes
```

(That message is terrible, but you get the idea)

**Comment by fvictorio** [comment]
Related: #1435

**Comment by fvictorio** [comment]
We just released Hardhat [2.4.0](https://github.com/nomiclabs/hardhat/releases/tag/hardhat-core-v2.4.0), which includes the `hardhat_setCode` RPC method.
